polygeist/tools/cgeist/Lib/CodeGenTypes.cc

```diff
@@ -1583,43 +1580,201 @@ mlir::Type CodeGenTypes::getMLIRType(clang::QualType QT, bool *ImplicitRef,
-                                        "for this target. Use with caution.\n";
-        }
-      });
-      return Builder.getF16Type();
-    }
-
-    if (auto *IT = dyn_cast<llvm::IntegerType>(Ty))
-      return Builder.getIntegerType(IT->getBitWidth());
+    llvm::Type *Ty = CGM.getTypes().ConvertType(QualType(T, 0));
+    return Builder.getIntegerType(cast<llvm::IntegerType>(Ty)->getBitWidth());
   }
 
+  if (T->isBuiltinType())
+    return getMLIRType(cast<clang::BuiltinType>(T));
+
   LLVM_DEBUG(llvm::dbgs() << "QT: "; QT->dump(); llvm::dbgs() << "\n");
   llvm_unreachable("unhandled type");
 }
 
+mlir::Type CodeGenTypes::getMLIRType(const clang::BuiltinType *BT) const {
+  assert(BT && "Expecting valid pointer");
+
+  mlir::OpBuilder Builder(TheModule->getContext());
+  mlir::LLVM::TypeFromLLVMIRTranslator TypeTranslator(*TheModule->getContext());
+
+  switch (BT->getKind()) {
+  case BuiltinType::Void:
+    return Builder.getNoneType();
+
+  case BuiltinType::ObjCId:
+  case BuiltinType::ObjCClass:
+  case BuiltinType::ObjCSel:
+    return Builder.getIntegerType(8);
+
+  case BuiltinType::Bool:
+    // TODO: boolean types should be represented as i1 rather than i8.
+    return Builder.getIntegerType(8);
+
+  case BuiltinType::Char_S:
+  case BuiltinType::Char_U:
+  case BuiltinType::SChar:
+  case BuiltinType::UChar:
+  case BuiltinType::Short:
+  case BuiltinType::UShort:
+  case BuiltinType::Int:
+  case BuiltinType::UInt:
+  case BuiltinType::Long:
+  case BuiltinType::ULong:
+  case BuiltinType::LongLong:
+  case BuiltinType::ULongLong:
+  case BuiltinType::WChar_S:
+  case BuiltinType::WChar_U:
+  case BuiltinType::Char8:
+  case BuiltinType::Char16:
+  case BuiltinType::Char32:
+  case BuiltinType::ShortAccum:
+  case BuiltinType::Accum:
+  case BuiltinType::LongAccum:
+  case BuiltinType::UShortAccum:
+  case BuiltinType::UAccum:
+  case BuiltinType::ULongAccum:
+  case BuiltinType::ShortFract:
+  case BuiltinType::Fract:
+  case BuiltinType::LongFract:
+  case BuiltinType::UShortFract:
+  case BuiltinType::UFract:
+  case BuiltinType::ULongFract:
+  case BuiltinType::SatShortAccum:
+  case BuiltinType::SatAccum:
+  case BuiltinType::SatLongAccum:
+  case BuiltinType::SatUShortAccum:
+  case BuiltinType::SatUAccum:
+  case BuiltinType::SatULongAccum:
+  case BuiltinType::SatShortFract:
+  case BuiltinType::SatFract:
+  case BuiltinType::SatLongFract:
+  case BuiltinType::SatUShortFract:
+  case BuiltinType::SatUFract:
+  case BuiltinType::SatULongFract:
+    return Builder.getIntegerType(Context.getTypeSize(BT));
+
+  case BuiltinType::Float16:
+  case BuiltinType::Half:
+  case BuiltinType::BFloat16:
+    CGEIST_WARNING({
+      if (CGM.getTarget().shouldEmitFloat16WithExcessPrecision()) {
+        llvm::WithColor::warning() << "Experimental usage of _Float16. Code "
+                                      "generated will be illegal "
+                                      "for this target. Use with caution.\n";
+      }
+    });
```

Comment:
I was going to put a PR to drop this warning, as we're already handling this issue (we expand before operating with this kind of types). You can do it here if you like.

Reply:
Sure will drop the warning.